tensorpack/graph_builder/distributed.py

```diff
@@ -277,27 +303,97 @@ def build(self, get_grad_fn, get_opt_fn):
         get_opt_fn = memoized(get_opt_fn)
         # Build the optimizer first, before entering any tower.
         # This makes sure that learning_rate is a global variable (what we expect)
-        get_opt_fn()    # TODO get_opt_fn called before main graph was built
+        get_opt_fn()  # TODO get_opt_fn called before main graph was built
 
         # Ngpu * Nvar * 2
-        grad_list = DataParallelBuilder.build_on_towers(
+        self.grad_list = DataParallelBuilder.build_on_towers(
             self.towers, get_grad_fn,
             devices=self.raw_devices,
             use_vs=[True] * len(self.towers))  # open vs at each tower
-        DataParallelBuilder._check_grad_list(grad_list)
-
-        avg_grads = aggregate_grads(
-            grad_list, colocation=False, devices=self.raw_devices)
-        with tf.device(self.param_server_device):
-            ps_var_grads = DistributedReplicatedBuilder._apply_shadow_vars(avg_grads)
-            var_update_ops = self._apply_gradients_and_copy(
-                get_opt_fn(), grad_list, ps_var_grads)
-            self._shadow_vars = [v for (__, v) in ps_var_grads]
-            self._shadow_model_vars = DistributedReplicatedBuilder._shadow_model_variables(self._shadow_vars)
+        DataParallelBuilder._check_grad_list(self.grad_list)
+
+        if self.aggregation_frequency > 1:
+            # Create the variables which will be used to store aggregated updates
+            for k in range(len(self.grad_list)):
+                if k >= len(self.gpu_shadow_vars):
+                    self.gpu_shadow_vars.append([])
+                    for idx, (grad, var) in enumerate(self.grad_list[k]):
+                        if idx >= len(self.gpu_shadow_vars[k]):
+                            with tf.device(self.raw_devices[k]), tf.variable_scope("aggregation_variables"):
+                                grad_aggregation_variable_name = "task%dtower%didx%d" % (self.task_index, k, idx)
```

Comment:
don't know if this is consistent with the rest of the code base or if we are Python 3.6+, but `f"task{self.task_index}tower{k}idx{idx}"` might be more readable.